apps/passport-server/src/services/generic-issuance/AutoIssuanceProvider.ts

```diff
@@ -0,0 +1,153 @@
+import { IPipelineConsumerDB } from "../../database/queries/pipelineConsumerDB";
+import { PretixAtom } from "./pipelines/PretixPipeline";
+
+export class AutoIssuanceProvider {
+  private pipelineId: string;
+  private autoIssuanceConfig: AutoIssuanceOptions[];
+
+  public constructor(
+    pipelineId: string,
+    autoIssuanceConfig: AutoIssuanceOptions[]
+  ) {
+    this.pipelineId = pipelineId;
+    this.autoIssuanceConfig = autoIssuanceConfig;
+  }
+
+  public async dripNewManualTickets(
+    consumerDB: IPipelineConsumerDB,
+    existingManualTickets: ManualTicket[],
+    realTickets: PretixAtom[]
+  ): Promise<ManualTicket[]> {
+    const allConsumers = await consumerDB.loadAll(this.pipelineId);
+    const newManualTickets: ManualTicket[] = [];
+
+    for (const consumer of allConsumers) {
+      newManualTickets.push(
+        ...(await this.maybeIssueForUser(
+          consumer.email,
+          existingManualTickets,
+          realTickets
+        ))
+      );
+    }
+
+    return newManualTickets;
+  }
+
+  public async maybeIssueForUser(
+    userEmail: string,
+    existingManualTickets: ManualTicket[],
+    realTickets: PretixAtom[]
+  ): Promise<ManualTicket[]> {
+    const userRealTickets = realTickets.filter((t) => t.email === userEmail);
+    const userManualTickets = existingManualTickets.filter(
+      (t) => t.attendeeEmail === userEmail
+    );
+    const newManualTickets: ManualTicket[] = [];
+
+    for (const autoIssuance of this.autoIssuanceConfig) {
+      const permissioningRealTicket = userRealTickets.find((t) =>
+        ticketMatchesCriteria(t, autoIssuance.memberCriteria)
+      );
+      const permissioningManualTicket = userManualTickets.find((t) =>
+        ticketMatchesCriteria(t, autoIssuance.memberCriteria)
+      );
+
+      const matchesIssuanceMembershipCriteria: boolean =
+        !!permissioningRealTicket || !!permissioningManualTicket;
+
+      if (!matchesIssuanceMembershipCriteria) {
+        continue;
+      }
+
+      if (
+        !canIssueInThisEpoch(autoIssuance, existingManualTickets, userEmail)
+      ) {
+        continue;
+      }
+
+      for (let i = 0; i < autoIssuance.quantity; i++) {
+        const newManualTicket: ManualTicket = {
+          attendeeEmail: userEmail,
+          attendeeName:
+            permissioningRealTicket?.name ??
+            permissioningManualTicket?.attendeeName ??
+            "no name",
+          eventId: autoIssuance.eventId,
+          productId: autoIssuance.productId,
+          id: randomUUID(),
+          timeCreated: new Date().toISOString()
+        };
+
+        newManualTickets.push(newManualTicket);
+      }
+    }
+
+    return newManualTickets;
+  }
+}
+
+export function anyTicketMatchesCriteria(
+  ts: Array<PretixAtom | ManualTicket>,
+  criterias: MemberCriteria[]
+): boolean {
+  return !!ts.find((t) => ticketMatchesCriteria(t, criterias));
+}
+
+export function ticketMatchesCriteria(
+  t: PretixAtom | ManualTicket,
```

Comment:
This is going to be a bit annoying on the Lemonade pipeline, as Lemonade atoms use different field names for eventId/productId. We probably ought to change the field names on Lemonade atoms to match these.

Reply:
Have done this in the #1663 branch

Reply:
I've also copied this function to a `utils.ts` file as I'm using the same interface to represent the kinds of tickets a user is capable of checking in for offline check-in purposes.